contents/security/sqlInjection.md

```diff
@@ -0,0 +1,149 @@
+to try to gain access, the constructured query would look like this:
+
+```sql
+SELECT * FROM Users WHERE Username = ‘Alice’ 
+AND Password = ‘password’
+```
+
+This query would look at the "Users" table in the database, and return any entries where the 
+column "Username" matches "Alice" and the column "Password" matches "password". 
+The intention may be to allow access to the website if the query returns any entry, as a result 
+would imply a matching username and password.
+
+
+However, are users limited to just filling in the parameters? Some websites directly use the input given, 
+thus the users are able to modify the query, or even to add in their own. This can easily be used 
+for malicious purposes. For example, users are able to add parameters to the query.
+
+```sql
+Username: foo
+Password: bar’ OR ‘’=’
+```
+
+The SQL command string built from this input would be as follows:
+
+```sql
+SELECT * FROM Users WHERE Username = ‘foo’ 
+AND Password = ‘bar’ OR ‘’=’’
+```
+
+This query will always return true, as ‘’=’’ is always true, and it is connected using an 
+OR comparator. As such, the query can be simplified to this:
+
+```sql
+SELECT * FROM Users
+```
+
+This will return all the rows from the Users table in the database, regardless of username 
+or password entered. If used in user authentication, the user will likely be able to gain
+access to the system.
+
+
+## Preventing SQL Injection ##
+
+There are a couple of ways to protect your website against SQL injection attacks. The two most
+common ways are:
+
+### Prepared Statements using Parameterized Queries ###
+
+By defining all the SQL code first and passing in the parameters afterwards, you can make 
+the database distinguish the difference between code and data. It would treat the values 
+entered by the user as a parameter, and would not allow it to alter the query being 
+executed. The way to achieve this varies by language, but it is easy to implement and effective.
+
+For example, instead of writing this in Java:
+
+```java
+String query = "SELECT account_balance FROM user_data WHERE user_name = " + request.getParameter("customerName");
+Statement statement = connection.createStatement();
+ResultSet results = statement.executeQuery( query );
+```
+
+You can prepare the statement like this:
+
+```java
+ String custname = request.getParameter("customerName");
+ String query = "SELECT account_balance FROM user_data WHERE user_name = ? ";
+ PreparedStatement pstmt = connection.prepareStatement( query );
+ pstmt.setString( 1, custname ); 
+ ResultSet results = pstmt.executeQuery( );
+```
+
+The "?" in the query string is a placeholder for a value. Since it is treated as a value to be used, 
+the database will not allow it to modify the query, and it would not be able to affect the structure 
+of the SQL statement. As such, the query will always execute the command as was intended.
```

Comment:
I would prefer your old example with 2 `?`, and then line number 4 in the code will use `custname` as a string value for the first occurence of `?`, and if you write `pstmt.setstring(2, ...)` thenit will replace the value for the 2nd occurence of `?`.

Basically I would like to see more detailed explanation  of the substitution line  of code,  and  also a slightly more sophisticated example to illustrate the use of `?`

Reply:
I changed the query so it uses the same one used in the examples earlier on. I'm not too sure how much space I should spend on explaining '<?>', because it's not used in every language, so it might not be too relevant rather than just discussing that it exists and helps.

Reply:
I think even if the other languages don't use `?`, they should use some similar constructs, so it is okay to explain the idea more clearly. My point is that the current explanation `The "?" in the query string is a placeholder for a value` is too short and people might not be able to understand. You can explain it more like "... is a placeholder for a string value. It is later substituted with the string value of `input_username` by the string value by the call `pstmt.setString(1, input_username)`"

Reply:
Ok, I see your point. Updated it now, thanks! =)